### PR TITLE
Handle recursive conversion in base type

### PR DESCRIPTION
**Pull request by fbrausse**
Turns out, I had a patch for #1776 laying around since a while. It's hard to get a good regression test as there seems to be a problem with accessing public members of a base class (1776-2 and 1776-3).

### CODE REVIEW

**Review by XLiZHI — APPROVED** [review]
LGTM

File: regression/esbmc-cpp/bug_fixes/1776-4/gh-1776.cc [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+};
+class b : public a {};
+
+int main()
+{
+	b B;
+	a *A = &B;
+	A->c = A;
+	assert(A->c == A->c->c);
+}
```
Comment (XLiZHI) [inline regression/esbmc-cpp/bug_fixes/1776-4/gh-1776.cc:17]: This TC can also pass in the current master.

Comment (fbrausse) [reply regression/esbmc-cpp/bug_fixes/1776-4/gh-1776.cc:17]: Hm, good point, meaning that this TC is pretty useless for this PR. Probably worth investigating the failures of the other ones. For reference, with this PR 1776-2 prints:
```
Generating GOTO Program
ERROR: Looking up index of nonexistant member "c" in struct/union "b"
Aborted
```
and 1776-3 does this:
```
[Counterexample]


State 1 file gh-1776.cc line 15 column 2 function main thread 0
----------------------------------------------------
Violated property:
  file gh-1776.cc line 15 column 2 function main
  dereference failure: Access to object out of bounds


VERIFICATION FAILED
```

**Comment by intrigus-lgtm** [comment]
The PR doesn't handle this test case:

```cpp
struct a;
struct b;
struct c
{
  void d(b);
};
struct a : c
{
  virtual ~a();
};
struct b : a
{
};
```

Crashes with (line number is not 100% accurate):
```shell
esbmc/src/clang-cpp-frontend/clang_cpp_convert_vft.cpp:561: void clang_cpp_convertert::add_vtable_variable_symbols(const clang::CXXRecordDecl&, const struct_typet&, const switch_table&): Assertion `vt_symb_type' failed.
```

While this test case works fine (note that `b` does not extend `a` anymore):
```cpp
struct a;
struct b;
struct c
{
  void d(b);
};
struct a : c
{
  virtual ~a();
};
struct b /* : a */
{
};
```

**Comment by intrigus-lgtm** [comment]
Thanks for creating the test case commit, which I've applied to my own PR: #1781.

**Comment by fbrausse** [comment]
Superseded by #1781. Also: doesn't make a lot of sense deriving from an incomplete type. It's better to keep to the C/C++ parsing rules.

**Pull request closed without merging**